### ISSUE TITLE
Clash over-eagerly reports "dubious primitive instantiations"

### ISSUE THREAD

**Issue by alex-mckenna** [clash-lang/clash-compiler#2046]
Since https://github.com/clash-lang/clash-compiler/commit/78852675608b9878cb18be560940d7c43805f474, Clash has been more eager to report when designs include primitives on types like `Integer` and `Natural`. These warnings are desirable when such operations remain in the output, but Clash can be too eager to show them to users (see #1918). Before Clash 1.6 we should improve the quality of this reporting to only show such warnings when they are strictly necessary (i.e. when the primitives are contained in the result of normalization).

Designs like this one (from #1918) should not display the "dubious primitive instantiation" warning, as the primitive is not rendered in the resulting HDL.

```haskell
module T2046 where

import Clash.Prelude

{-# NOINLINE topEntity #-}
topEntity
  :: Vec 5 Int
  -> Index 5
  -> Int
topEntity = (!!)
```

**Comment by alex-mckenna** [comment]
Suggested fix (from @christiaanb): make the `fromEnum` functions on Clash's numeric types primitives if they go through `Integer` / `Natural` so the warning is suppressed
